### ISSUE TITLE
[Usage]: 输出的质量问题

### ISSUE THREAD

**Issue by zzc98** [yhfgyyf/vllm-deepseek-v4-sm89#52]
按照此项目能部署起来，在 AutomationBench Public 评测只有7分。在H20上能达到24（官方给的数据是25.1）。差的有点多

**Comment by lightspear** [comment]
是的，我发现很多时候格式也输出有问题的

<img width="535" height="129" alt="Image" src="https://github.com/user-attachments/assets/07602d64-d27b-4645-ae39-41b45684cf70" />

**Comment by yhfgyyf** [comment]
使用的哪个版本，是否安装了 PR https://github.com/yhfgyyf/vllm-deepseek-v4-sm89/pull/51.

**Comment by zzc98** [comment]
> 使用的哪个版本，是否安装了 PR [#51](https://github.com/yhfgyyf/vllm-deepseek-v4-sm89/pull/51).

vllm 0.23.1rc1，按照 PR51对 `/usr/local/lib/python3.12/dist-packages/vllm/models/deepseek_v4/nvidia/ops/sm12x_mqa.py` 这个文件进行了修改。

<img width="626" height="35" alt="Image" src="https://github.com/user-attachments/assets/f7a7f120-3eb3-4b38-8ee0-9a19503260af" />

<img width="398" height="130" alt="Image" src="https://github.com/user-attachments/assets/e5617bf8-ad5c-47d1-b100-db58047a4060" />

启动命令：
```bash
vllm serve /mnt/data/checkpoints/DeepSeek-V4-Flash-0731 --port 40061 --served-model-name deepseek-v4-flash-0731 --max-model-len=262144 --block-size=256 --enable-expert-parallel --tensor-parallel-size=8 --tokenizer-mode=deepseek_v4 --tool-call-parser=deepseek_v4 --enable-auto-tool-choice --reasoning-parser=deepseek_v4 --trust-remote-code --gpu-memory-utilization=0.75 --kv-cache-dtype=fp8_ds_mla --attention-backend=FLASHINFER_MLA_SPARSE_DSV4 --disable-access-log-for-endpoints /metrics --enable-prompt-tokens-details --speculative-config={"method":"dspark","num_speculative_tokens":6,"draft_sample_method":"greedy"} --max-num-seqs=8 --max-num-batched-tokens=2048
```

**Comment by lightspear** [comment]
sm12x_mqa.py 是的我也是修复过这个，依然有问题的，同上

**Comment by nieziqi123** [comment]
我也出现了，PR-51 sm12x_mqa.py已修复

<img width="527" height="445" alt="Image" src="https://github.com/user-attachments/assets/c4b167b6-f7e9-4461-8bb1-5bbd01fa6307" />

**Comment by zzc98** [comment]
> 我也出现了，PR-51 sm12x_mqa.py已修复
> 
> <img alt="Image" width="527" height="445" src="https://private-user-images.githubusercontent.com/51815799/633502182-c4b167b6-f7e9-4461-8bb1-5bbd01fa6307.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3ODYzNTA0ODMsIm5iZiI6MTc4NjM1MDE4MywicGF0aCI6Ii81MTgxNTc5OS82MzM1MDIxODItYzRiMTY3YjYtZjdlOS00NDYxLThiYjEtNWJiZDAxZmE2MzA3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjA4MTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwODEwVDA4MjMwM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQzOTFlYWNkYzIxZGJmOWU5OTZjODY3NzI5NDZmNTRhNzU1MmJmYTdhNGIwZjMxNjI1NGRlYmIxOGM3N2E0YjkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JnJlc3BvbnNlLWNvbnRlbnQtdHlwZT1pbWFnZSUyRnBuZyJ9.eIzjxJwbehj5tUgnre88m0Lke69Hy26GT-XtwH7hf3c">

看提示词，正常来说，工具调用应该是这种格式，难道是 DSML 这个特殊token的问题？
```text
<｜DSML｜tool_calls>
<｜DSML｜invoke name="$TOOL_NAME">
<｜DSML｜parameter name="$PARAMETER_NAME" string="true|false">$PARAMETER_VALUE</｜DSML｜parameter>
</｜DSML｜invoke>
</｜DSML｜tool_calls>
```

**Comment by JCRPaquin** [comment]
I ran a comparison between vLLM (with PR-51) on 8xL40S TP=8 and API provider logits. The local vLLM instance seems to be generating a top-20 distribution within tolerance of the cloud providers.

| pair | prompts | aligned pos | non-EOS | top-1 agree | top-20 overlap | median JSD | median mean \|_logprob\| | ref argmax outside local top-20 |
|---|---|---:|---:|---:|---:|---:|---:|---:|
| local vs Cloudflare | 36 | 223 | 223 | 95.5% | 89.0% | 0.0000 | 0.803 | 0.0% |
| local vs CoreWeave | 36 | 243 | 235 | 92.6% | 91.8% | 0.0000 | 0.449 | 0.0% |
| Cloudflare vs CoreWeave | 36 | 219 | 219 | 96.3% | 89.6% | 0.0000 | 0.678 | 0.0% |

(non-EOS = samples where neither provider sampled an end-of-sentence token)

I also had an agent run a comparison between a torch implementation of sparse attention vs flashinfer within a single layer. Everything came back within FP8 dequantization tolerance; errors are likely quantization noise and the kernels are implemented/compiled correctly.

**Comment by zzc98** [comment]
> I ran a comparison between vLLM (with PR-51) on 8xL40S TP=8 and API provider logits. The local vLLM instance seems to be generating a top-20 distribution within tolerance of the cloud providers.
> 
> pair	prompts	aligned pos	non-EOS	top-1 agree	top-20 overlap	median JSD	median mean |_logprob|	ref argmax outside local top-20
> local vs Cloudflare	36	223	223	95.5%	89.0%	0.0000	0.803	0.0%
> local vs CoreWeave	36	243	235	92.6%	91.8%	0.0000	0.449	0.0%
> Cloudflare vs CoreWeave	36	219	219	96.3%	89.6%	0.0000	0.678	0.0%
> (non-EOS = samples where neither provider sampled an end-of-sentence token)
> 
> I also had an agent run a comparison between a torch implementation of sparse attention vs flashinfer within a single layer. Everything came back within FP8 dequantization tolerance; errors are likely quantization noise and the kernels are implemented/compiled correctly.

Could you share your vLLM launch command

**Comment by JCRPaquin** [comment]
> Could you share your vLLM launch command

```
vllm serve /models/DeepSeek-V4-Flash-0731 --host 0.0.0.0 --port 8000 \
  --served-model-name deepseek-v4-flash --tensor-parallel-size 8 --gpu-memory-utilization 0.92 \
  --max-model-len 262144 --max-num-seqs 4 --block-size 256 --kv-cache-dtype fp8_ds_mla \
  --attention-backend FLASHINFER_MLA_SPARSE_DSV4 --reasoning-parser deepseek_v4 \
  --tool-call-parser deepseek_v4 --max-num-batched-tokens 2048 --enable-auto-tool-choice \
  --trust-remote-code --compilation-config {"cudagraph_mode":"NONE"} \
  --speculative-config {"method":"dspark","num_speculative_tokens":2,"draft_sample_method":"greedy"}
```

I apologize, I forgot to set the graph mode to FULL_AND_PIECEWISE. I'll re-run the tests at some point in the next couple of days with CUDA graph enabled.

**Comment by Zephyr69** [comment]
> I ran a comparison between vLLM (with PR-51) on 8xL40S TP=8 and API provider logits. The local vLLM instance seems to be generating a top-20 distribution within tolerance of the cloud providers.
> 
> pair	prompts	aligned pos	non-EOS	top-1 agree	top-20 overlap	median JSD	median mean |_logprob|	ref argmax outside local top-20
> local vs Cloudflare	36	223	223	95.5%	89.0%	0.0000	0.803	0.0%
> local vs CoreWeave	36	243	235	92.6%	91.8%	0.0000	0.449	0.0%
> Cloudflare vs CoreWeave	36	219	219	96.3%	89.6%	0.0000	0.678	0.0%
> (non-EOS = samples where neither provider sampled an end-of-sentence token)
> 
> I also had an agent run a comparison between a torch implementation of sparse attention vs flashinfer within a single layer. Everything came back within FP8 dequantization tolerance; errors are likely quantization noise and the kernels are implemented/compiled correctly.

https://github.com/wtdcode/vllm-backport This repo seems to work well for SM89. Could you also test its output correctness?

**Comment by JCRPaquin** [comment]
> https://github.com/wtdcode/vllm-backport This repo seems to work well for SM89. Could you also test its output correctness?

I'll add it to the list, but I'll likely need to wait until the weekend to test.

Interesting suggestion from that repo: `--hf-overrides '{"head_dtype": "float32"}'`
> Tips:
> - head dtype override helps reduce garbage outputs.

**Comment by zzc98** [comment]
> https://github.com/wtdcode/vllm-backport This repo seems to work well for SM89. Could you also test its output correctness?

I used this repository and scored 24 points on the AutomationBench Public evaluation. This shows that the code is fine.

**Comment by JCRPaquin** [comment]
> --compilation-config {"cudagraph_mode":"NONE"}

Changing this to `FULL_AND_PIECEWISE` and re-running the comparisons with Cloudflare and CoreWeave returned effectively the same results. Not sure where else an output quality problem might originate from.

> --hf-overrides '{"head_dtype": "float32"}'

This didn't change the metrics in a meaningful way, either.
